pkg/pipeline/write/write_loki.go

```diff
@@ -0,0 +1,283 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package write
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/netobserv/flowlogs2metrics/pkg/api"
+	pUtils "github.com/netobserv/flowlogs2metrics/pkg/pipeline/utils"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/netobserv/flowlogs2metrics/pkg/config"
+
+	logAdapter "github.com/go-kit/kit/log/logrus"
+	jsonIter "github.com/json-iterator/go"
+	"github.com/netobserv/loki-client-go/loki"
+	"github.com/netobserv/loki-client-go/pkg/backoff"
+	"github.com/netobserv/loki-client-go/pkg/urlutil"
+	"github.com/prometheus/common/model"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	keyReplacer = strings.NewReplacer("/", "_", ".", "_", "-", "_")
+)
+
+type emitter interface {
+	Handle(labels model.LabelSet, timestamp time.Time, record string) error
+}
+
+const channelSize = 1000
+
+// Loki record writer
+type Loki struct {
+	lokiConfig loki.Config
+	apiConfig  api.WriteLoki
+	client     emitter
+	timeNow    func() time.Time
+	ready      bool
+	in         chan config.GenericMap
+	exitChan   chan bool
+}
+
+func (l *Loki) IsReady() bool {
+	return l.ready
+}
+
+func buildLokiConfig(c *api.WriteLoki) (loki.Config, error) {
+	batchWait, err := time.ParseDuration(c.BatchWait)
+	if err != nil {
+		return loki.Config{}, fmt.Errorf("failed in parsing BatchWait : %v", err)
+	}
+
+	timeout, err := time.ParseDuration(c.Timeout)
+	if err != nil {
+		return loki.Config{}, fmt.Errorf("failed in parsing Timeout : %v", err)
+	}
+
+	minBackoff, err := time.ParseDuration(c.MinBackoff)
+	if err != nil {
+		return loki.Config{}, fmt.Errorf("failed in parsing MinBackoff : %v", err)
+	}
+
+	maxBackoff, err := time.ParseDuration(c.MaxBackoff)
+	if err != nil {
+		return loki.Config{}, fmt.Errorf("failed in parsing MaxBackoff : %v", err)
+	}
+
+	cfg := loki.Config{
+		TenantID:  c.TenantID,
+		BatchWait: batchWait,
+		BatchSize: c.BatchSize,
+		Timeout:   timeout,
+		BackoffConfig: backoff.BackoffConfig{
+			MinBackoff: minBackoff,
+			MaxBackoff: maxBackoff,
+			MaxRetries: c.MaxRetries,
+		},
+		Client: c.ClientConfig,
+	}
+	var clientURL urlutil.URLValue
+	err = clientURL.Set(strings.TrimSuffix(c.URL, "/") + "/loki/api/v1/push")
+	if err != nil {
+		return cfg, fmt.Errorf("failed to parse client URL: %w", err)
+	}
+	cfg.URL = clientURL
+	return cfg, nil
+}
+
+func (l *Loki) ProcessRecord(record config.GenericMap) error {
+	if !l.IsReady() {
+		return errors.New("WriteLoki is not ready")
+	}
```

Comment:
It seems you can remove this; `IsReady` function and `ready` field are always `true` in your implementation

Reply:
➕1

Reply:
removed, done